scripts/install-build-tools.sh

```diff
@@ -14,35 +15,208 @@ if (( $EUID != 0 )); then
+    else
+      echo "Installed python${PY_VERSION}-venv"
+    fi
+  fi
+  # create virtual environment with specific python version
+  if "${PY_LOC}" -m venv "${ENV_NAME}"; then
+    echo "Virtual environment '${ENV_NAME}' created"
+  else
+    echo "Virtual environment creation failed"; exit 1
+  fi
+  # activate virtual environment
+  source "${ROOT}/scripts/activate-venv.sh"
+  # install python packages
+  if "${ROOT}/${ENV_NAME}/bin/pip" install -r requirements.txt; then
+    echo "Success installing python packages"
+  else
+    echo "Failure installing python packages"
+    deactivate
+    exit 1
+  fi
+  deactivate
+}
+
+
+echo "OS Type: $OSTYPE"
+# macOS install with homebrew
 if [[ "$OSTYPE" == "darwin"* ]]; then
+
+  # macOS does not support running shell scripts as root with homebrew
+  if [ $EUID -eq 0 ]; then
+    echo -e "Mac users should run this script without 'sudo'. Exiting..."
+    exit 1
+  fi
+
   CPUS=$(sysctl -n hw.ncpu)
   # ensure development environment is set correctly for clang
   $SUDO xcode-select -switch /Library/Developer/CommandLineTools
-  brew install llvm@14 googletest google-benchmark lcov make wget cmake curl
+
+  if ! brew --version &>/dev/null; then
+    echo -e "${cyan}Homebrew is required to install dependencies.${end}"
+    exit 1
+  fi
+
+  brew install llvm@14 googletest google-benchmark lcov make wget cmake bash bc
+  brew upgrade bash
+
+  BREW_ROOT=$(brew --prefix)
+
   CLANG_TIDY=/usr/local/bin/clang-tidy
   if [ ! -L "$CLANG_TIDY" ]; then
-    $SUDO ln -s $(brew --prefix)/opt/llvm@14/bin/clang-tidy /usr/local/bin/clang-tidy
+    $SUDO ln -s "${BREW_ROOT}/opt/llvm@14/bin/clang-tidy" /usr/local/bin/clang-tidy
   fi
   GMAKE=/usr/local/bin/gmake
   if [ ! -L "$GMAKE" ]; then
     $SUDO ln -s $(xcode-select -p)/usr/bin/gnumake /usr/local/bin/gmake
   fi
-fi
 
-if [[ "$OSTYPE" == "linux-gnu"* ]]; then
-  $SUDO apt update
-  $SUDO apt install -y build-essential wget cmake libgtest-dev libbenchmark-dev lcov git software-properties-common rsync unzip
+  # check if supported version of python3 is already installed, and save the version
+  for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+    if which python${PY_VERS}; then
+      # save the installed version if found
+      PY_INSTALLED=${PY_VERS}
+      break
+    fi
+  done
+
+  # install valid python version if not installed yet
+  if [ -z "$PY_INSTALLED" ]; then
+    for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+      FULL_PY="python${PY_VERS}"
+      # try to install python version from homebrew and verify installation
+      if brew install "${FULL_PY}"; then
+        echo "${FULL_PY} installed successfully"
+        PY_INSTALLED=${PY_VERS}
+        break
+      else
+        # don't exit as we can try on another version if not the last one
+        echo "${FULL_PY} installation failed"
+      fi
+    done
+  fi
+
+  # leave if no valid python version is installed
+  if ! $(which python${PY_INSTALLED}); then
```

Comment:
This probably doesn't do what you expect. This is negating the string printed by `which python{$PY_INSTALLED}`.

Note that because the full per-os blocks repeat near-identical steps, there's another instance of this below.

Instead, you likely want:
```suggestion
  if ! which "python${PY_INSTALLED}" &> /dev/null; then
```

Reply:
understood, thanks